lib/rspec-puppet-facts.rb

```diff
@@ -95,28 +94,46 @@ def on_supported_os(opts = {})
       else
         opts[:hardwaremodels].each do |hardwaremodel|
           filter << {
-              :facterversion   => facter_version_filter,
               :operatingsystem => os_sup['operatingsystem'],
               :hardwaremodel   => hardwaremodel,
           }
         end
       end
     end
 
+    # FacterDB may have newer versions of facter data for which it contains a subset of all possible
+    # facter data (see FacterDB 0.5.2 for Facter releases 3.8 and 3.9). In this situation we need to
+    # cycle through and downgrade Facter versions per platform type until we find matching Facter data.
+    filter.each do |filter_spec|
+      facter_version_filter = RspecPuppetFacts.facter_version_to_filter(facterversion)
+      db = FacterDB.get_facts(filter_spec.merge({ :facterversion =>  facter_version_filter }))
+
+      version = facterversion
+      while db.empty? && version !~ /\d+\.0($|\.\d+)/
```

Comment:
This loop can end up hanging forever.  `down_facter_version` appears to keep decrementing past 0 into versions with negative numbers.
```
"3.9"
"3.8."
"3.7."
"3.6."
"3.5."
"3.4."
"3.3."
"3.2."
"3.1."
"3.0."
"3.-1."
"3.-2."
"3.-3."
"3.-4."
"3.-5."
...
```